design/src/rfcs/rfc0017_logging_sensitive.md

```diff
@@ -0,0 +1,376 @@
+];
+let router = aws_smithy_http_server::routing::Router::new_rest_json_router(routes);
+```
+
+As a request enters this layer it should record the HTTP headers, status code, and URI.
+
+The following model
+
+```smithy
+@readonly
+@http(uri: "/inventory/{name}", method: "GET")
+operation Inventory {
+    input: Product,
+    output: Stocked
+}
+
+@input
+structure Product {
+    @required
+    @sensitive
+    @httpLabel
+    name: String
+}
+
+@output
+structure Stocked {
+    @sensitive
+    @httpResponseCode
+    code: String,
+}
+```
+
+should generate the following
+
+```rust
+// NOTE: This code is intended to show behavior - it does not compile
+
+const SENSITIVE_MARKER: &str = "_redacted_";
+
+pub struct InventoryLogging<S> {
+    inner: S
+}
+
+impl<S> InventoryLogging<S> {
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner
+        }
+    }
+}
+
+impl<B, S> Service<Request<B>> for InventoryLogging<S>
+where
+    S: Service<Request<B>>
+{
+    type Response = Response<BoxBody>;
+    type Error = S::Error;
+    type Future = /* Implementation detail */;
+
+    fn call(&mut self, request: Request<B>) -> Self::Future {
+        // Remove sensitive data from path
+        let mut uri = redact(request.uri());
+
+        let fut = async {
+            let response = self.inner.call(request).await;
+
+            debug!(status_code = %Sensitive(request.status()), "sending response");
+
+            response
+        };
+
+        // Instrument the future with a span
+        let span = span!(%method, %uri, headers = request.headers(), "received request");
+        fut.instrument(span)
+    }
+}
+```
+
+As this path is latency-sensitive, careful implementation is required to avoid excess allocations during redaction of sensitive data. Wrapping `Uri` and `HeaderMap` then providing a new `Display` implementation which skips over the sensitive data is preferable over allocating a new `String`/`HeaderMap` and mutating it.
+
+### Internal Guideline
+
+A guideline should be made available to internal smithy developers to outline the following:
+
+- The [HTTP bindings traits](#http-binding-traits) and why they are of concern.
+- The [Debug implementation](https://github.com/awslabs/smithy-rs/pull/229) on structures.
+- How to use the `Sensitive` struct and the `debug-logging` feature flag described in [Debug Logging](#debug-logging).
+
+### Public Guideline
+
+A guideline should be made available to customers to outline the following:
+
+- The [HTTP bindings traits](#http-binding-traits) and why they are of concern.
+- Warn against the two potential leaks described in [Scope and Guidelines](#scope-and-guidelines):
+  - Sensitive data leaking from third-party dependencies.
+  - Sensitive data leaking from middleware applied to the `Router`.
+- How to use the `Sensitive` struct and the `debug-logging` feature flag described in [Debug Logging](#debug-logging).
+
+## Alternative Proposals
```

Comment:
I think I lean towards the main solution you proposed as the code-generation allows us to precisely tailor the logging on a per-operation basis, while using runtime switches feels more complicated and we could probably loose control without noticing in the future.